components/profile/trace-dump.js

```diff
@@ -0,0 +1,504 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*** State *******************************************************************/
+
+window.COLORS = [
+  "#0088cc",
+  "#5b5fff",
+  "#b82ee5",
+  "#ed2655",
+  "#f13c00",
+  "#d97e00",
+  "#2cbb0f",
+  "#0072ab",
+];
+
+window.MIN_TRACE_TIME = 100000; // .1 ms
+
+// A class containing the cleaned up trace state.
+window.State = (function () {
+  return class {
+    constructor() {
+      // The traces themselves.
+      this.traces = null;
+
+      // Maximimum and minimum times seen in traces. These get normalized to be
+      // relative to 0, so after initialization minTime is always 0.
+      this.minTime = Infinity;
+      this.maxTime = 0;
+
+      // The current start and end of the viewport selection.
+      this.startSelection = 0;
+      this.endSelection = 0;
+
+      // The current width of the window.
+      this.windowWidth = window.innerWidth;
+
+      // Whether the user is actively grabbing the left or right grabby, or the
+      // viewport slider.
+      this.grabbingLeft = false;
+      this.grabbingRight = false;
+      this.grabbingSlider = false;
+
+      // Maps category labels to a persistent color so that they are always
+      // rendered the same color.
+      this.colorIndex = 0;
+      this.categoryToColor = Object.create(null);
+
+      this.initialize();
+    }
+
+    // Clean up and massage the trace data.
+    initialize() {
+      this.traces = TRACES.filter(t => t.endTime - t.startTime >= MIN_TRACE_TIME);
+      window.TRACES = null;
+
+      this.traces.sort((t1, t2) => {
+        let cmp = t1.startTime - t2.startTime;
+        if (cmp !== 0) {
+          return cmp;
+        }
+
+        return t1.endTime - t2.endTime;
+      });
+
+      this.findMinTime();
+      this.normalizeTimes();
+      this.removeIdleTime();
+      this.findMaxTime();
+
+      this.startSelection = 3 * this.maxTime / 8;
+      this.endSelection = 5 * this.maxTime / 8;
+    }
+
+    // Find the minimum timestamp.
+    findMinTime() {
+      this.minTime = this.traces.reduce((min, t) => Math.min(min, t.startTime),
+                                        Infinity);
+    }
+
+    // Find the maximum timestamp.
+    findMaxTime() {
+      this.maxTime = this.traces.reduce((max, t) => Math.max(max, t.endTime),
+                                        0);
+    }
+
+    // Normalize all times to be relative to the minTime and then reset the
+    // minTime to 0.
+    normalizeTimes() {
+      for (let i = 0; i < this.traces.length; i++) {
+        let trace = this.traces[i];
+        trace.startTime -= this.minTime;
+        trace.endTime -= this.minTime;
+      }
+      this.minTime = 0;
+    }
+
+    // Remove idle time between traces. It isn't useful to see and makes
+    // visualizing the data more difficult.
+    removeIdleTime() {
```

Comment:
> It would be nice if this would leave a marker so you can tell where the breaks are. Also, it may be useful to be able to toggle this on or off.

I found that the vast majority of time was idle, presumably waiting on the network, which is not yet traced by the profiler, and it was sort of hard to find interesting things. This could just as easily be interpreted as reason to trace more things :) We could definitely make it toggle-able or insert placeholders to show that there was idle time squashed, but I'd prefer to get the initial thing landed before futzing about with details too much.